### PR TITLE
cli: config: add support for --show-origin

### LINKED ISSUE #5119: config: add support for --show-origin
Same as `git config`:
```
$ git config --list --show-origin
file:/home/efiop/.gitconfig     user.email=ruslan@iterative.ai
file:/home/efiop/.gitconfig     user.name=Ruslan Kuprieiev
file:.git/config        core.repositoryformatversion=0
file:.git/config        core.filemode=true
file:.git/config        core.bare=false
file:.git/config        core.logallrefupdates=true
file:.git/config        remote.origin.url=git@github.com:efiop/dvc
file:.git/config        remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config        branch.master.remote=origin
file:.git/config        branch.master.merge=refs/heads/master
file:.git/config        remote.upstream.url=git@github.com:iterative/dvc
file:.git/config        remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config        branch.fix-4050.remote=origin
file:.git/config        branch.fix-4050.merge=refs/heads/fix-4050
file:.git/config        branch.fix-4162.remote=origin
file:.git/config        branch.fix-4162.merge=refs/heads/fix-4162
```

i'm not personally sold on `file:` prefix in path (just because I don't really know why git does that, maybe there is a good reason), but the rest looks pretty good. We now have `dvc config --list`, so now we need `dvc config --list --show-origin` as the next step.

This will be realy handy to have for users to figure out where the configs are and also handy for us, developers, when debugging particular issues.

### PR DESCRIPTION
**Pull request by mrstegeman**
Support the --show-origin option for config, which, similar to git,
prefixes each config option with the source file it originated from.

Fixes #5119

* [X] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [X] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by shcheklein** [comment]
@mrstegeman hey! 👋 Thanks for the PR! I see that you set the docs checkbox, but I don't see any issues/PRs in the docs repo? Is it still WIP?

**Comment by mrstegeman** [comment]
Ah, apologies. I was thinking it was just documentation in this repo.

https://github.com/iterative/dvc.org/pull/2028

**Comment by shcheklein** [comment]
@mrstegeman no worries at all (the layout is not very usual indeed when you have a separate repo with docs). Thanks for the PR, we'll take a look asap!! Give us please a bit of time for this.

File: dvc/command/config.py [hunk 1]
```diff
@@ -90,9 +108,22 @@ def _check(self, conf, remote, section, opt=None):
...
+    @staticmethod
+    def _config_file_prefix(show_origin, config, level):
+        if not show_origin:
+            return ""
+
+        fname = config.files[level]
+
+        if level in ["local", "repo"]:
+            common = os.path.commonprefix([fname, Repo.find_root()])
+            fname = fname[len(common) + 1 :]
```
Comment (efiop) [inline dvc/command/config.py:124]: Maybe just relpath? Or am I missing something?

**Review by efiop — COMMENTED** [review]
Hi @mrstegeman !

Thanks for the PR! :pray: Looks good!

I totally forgot that we still didn't introduce `--repo` flag for `dvc config` and didn't switch to loading the merged config in `dvc config` without --repo/--local/--global/--system :man_facepalming: :man_facepalming: :man_facepalming: It is on our 2.0 todo list though. Which would make the task a bit harder :)

File: dvc/command/config.py (hunk 1 above)
Comment (mrstegeman) [reply dvc/command/config.py]: Yes, that is certainly more clear. Done.

**Comment by mrstegeman** [comment]
> I totally forgot that we still didn't introduce `--repo` flag for `dvc config` and didn't switch to loading the merged config in `dvc config` without --repo/--local/--global/--system 🤦‍♂️ 🤦‍♂️ 🤦‍♂️ It is on our 2.0 todo list though. Which would make the task a bit harder :)

Yep, that was the biggest difference I noticed versus `git config --list`. It shouldn't make the logic much more difficult, though.

**Comment by efiop** [comment]
> Yep, that was the biggest difference I noticed versus git config --list. It shouldn't make the logic much more difficult, though.

@mribeirodantas Would you be willing to introduce the changes for that? Our master branch is currently preparing for 2.0, so we can do backward-incompatible changes without any problems.

**Comment by mrstegeman** [comment]
@efiop The two changes are really independent of each other, but should be done separately. It probably makes the most sense to just merge this as is, and then figure out the logic to merge and print all of the configs in a separate PR. I'd be happy to take on the other PR as well, after this is merged.

**Comment by mrstegeman** [comment]
For my own reference: [2.0 checklist](https://github.com/iterative/dvc/issues/4841)

**Review by efiop — COMMENTED** [review]
@mrstegeman Sounds good!

Ok, this PR looks good, but we need to add a test that tests the output. Please see `tests/func/test_config.py::test_config_remote` for an example of using caplog.

File: dvc/command/config.py [hunk 2]
```diff
@@ -54,10 +68,14 @@ def run(self):
...
             conf = self.config.load_one(self.args.level)
+            prefix = self._config_file_prefix(
+                self.args.show_origin, self.config, self.args.level
+            )
+
             if remote:
                 conf = conf["remote"]
             self._check(conf, remote, section, opt)
-            logger.info(conf[section][opt])
+            logger.info("{}{}".format(prefix, conf[section][opt]))
```
Comment (efiop) [inline dvc/command/config.py:78]: `git config` won't care about `--show-origin` in this scenario, right?

File: tests/func/test_config.py [hunk 3]
```diff
@@ -48,9 +48,15 @@ def _do_test(self, local=False):
         self.assertEqual(ret, 0)
         self.assertTrue(self._contains(section, field, value, local))
 
+        ret = main(base + [section_field, value, "--show-origin"])
```
Comment (karajan1001) [inline tests/func/test_config.py:51]: Can we parametrize these tests?

**Comment by karajan1001** [comment]
> I totally forgot that we still didn't introduce `--repo` flag for `dvc config` and didn't switch to loading the merged config in `dvc config` without --repo/--local/--global/--system 🤦‍♂️ 🤦‍♂️ 🤦‍♂️ It is on our 2.0 todo list though. Which would make the task a bit harder :)

@efiop
Excuse me, any details for the coming changes in 2.0?  Currently `dvc config --list` and `dvc config` behave differently in default ( without --repo/--local/--global/--system)?  `dvc config`  only changes `--repo` config while `dvc config --list` shows a merged one?

**Comment by efiop** [comment]
> > I totally forgot that we still didn't introduce `--repo` flag for `dvc config` and didn't switch to loading the merged config in `dvc config` without --repo/--local/--global/--system man_facepalming man_facepalming man_facepalming It is on our 2.0 todo list though. Which would make the task a bit harder :)
> 
> @efiop
> Excuse me, any details for the coming changes in 2.0? Currently `dvc config --list` and `dvc config` behave differently in default ( without --repo/--local/--global/--system)? `dvc config` only changes `--repo` config while `dvc config --list` shows a merged one?

No, `dvc config --list` and `dvc config`(in read mode) behave the same by default, but the problem is that they show `repo` level config by default, and not the merged config. But if you try to write something with it - it writes to repo config by default, which is correct. So in read mode we should use merged config and in write mode - continue using repo config by default. This is the same behavior as `git config` has.

File: dvc/command/config.py (hunk 2 above)
Comment (mrstegeman) [reply dvc/command/config.py:78]: I don't see why it would.

**Comment by mrstegeman** [comment]
Apologies for the delay. I added a couple tests for the output.

File: dvc/command/config.py (hunk 2 above)
Comment (efiop) [reply dvc/command/config.py:78]: @mrstegeman So should we care about it?

File: tests/func/test_config.py (hunk 3 above)
Comment (mrstegeman) [reply tests/func/test_config.py:51]: The tests are being reworked in #5179.

File: dvc/command/config.py (hunk 2 above)
Comment (mrstegeman) [reply dvc/command/config.py:78]: Oh, sorry, I misunderstood your question. Yes, git does use `--show-origin` when `--list` is not provided:

```
$ git config --show-origin branch.master.remote
file:.git/config	origin
```

Comment (efiop) [reply dvc/command/config.py:78]: Makes sense. Thanks for clarifying!

File: tests/func/test_config.py [hunk 4]
```diff
@@ -174,3 +186,28 @@ def test_config_remote(tmp_dir, dvc, caplog):
     caplog.clear()
     assert main(["config", "remote.myremote.region"]) == 0
     assert "myregion" in caplog.text
+
+
+def test_config_list(tmp_dir, dvc, caplog):
```
Comment (efiop) [inline tests/func/test_config.py:191]:
```suggestion
def test_config_show_origin(tmp_dir, dvc, caplog):
```

**Review by efiop — APPROVED** [review]
Thank you!

**Comment by efiop** [comment]
@mrstegeman I've added --repo flag and changed dvc config behaviour in #5179 (we need it for the upcoming 2.0 soon anyway), but temporarily forbid the use of --show-origin without --system/--global/--repo/--local flag so you could continue your work on a proper implementation as we've discussed here before. Please feel free to send a PR for that :slightly_smiling_face:

**Comment by mrstegeman** [comment]
@efiop Done, see #5188

**Pull request merged**